src/e3/anod/qualifiers_manager.py

```diff
@@ -0,0 +1,799 @@
+
+                # If choices exist then the qualifier value is constrained.
+                if qualifier["choices"] is not None:
+                    if qualifier_value not in qualifier["choices"]:
+                        self.__error(
+                            f"The {qualifier_name} qualifier value must be in "
+                            f'{qualifier["choices"]}. Got {qualifier_value}'
+                        )
+            elif qualifier["type"] == self.__TAG_QUALIFIER:
+                # The tag qualifier cannot be passed with a value
+                if qualifier_value:
+                    self.__error(
+                        f"The {qualifier_name} qualifier is a tag and does not expect "
+                        f'any values. Got "{qualifier_value}"'
+                    )
+            else:
+                raise AnodError(
+                    "An expected qualifier type was encountered during parsing "
+                    f'Got "{qualifier["type"]}"'
+                )
+
+        # Make sure all the declared qualifiers are stated
+        for qualifier_name, qualifier in self.qualifier_decls.items():
+            if qualifier["type"] == self.__KEY_VALUE_QUALIFIER:
+                if qualifier_name not in qualifier_configuration:
+                    # The qualifier must have a default otherwise it would not have
+                    # any values.
+                    if qualifier["default"] is None:
+                        self.__error(
+                            f"The {qualifier_name} qualifier was declared without a "
+                            "default value but not passed"
+                        )
+            elif qualifier["type"] != self.__TAG_QUALIFIER:
+                raise AnodError(
+                    "An expected qualifier type was encountered during parsing "
+                    f'Got "{qualifier["type"]}"'
+                )
+
+    def __key(
+        self, qualifier_configuration: dict[str, str]
+    ) -> tuple[tuple[str, str], ...]:
+        """Transform a parsed qualifier dict into a hashable object.
+
+        Ensure the returned key is unique and thus add the default
+        value of the qualifiers. This theoretically adds no information but
+        makes sure that implicit and explicit default values are handled the same way.
+
+        Cannot be called before the end of the declaration phase.
+
+        :param qualifier_configuration: A dictionnay of qualifier. The keys are the
+            qualifier names and the values are the coresponding qualifier values.
+        :return: A tuple containing the same information as qualifier_configuration.
+        """
+        if not self.is_declaration_phase_finished:
+            raise AnodError(
+                "The function 'key' cannot be used before the end of the declaration "
+                "phase"
+            )
+
+        # Set the key_value qualifier default values.
+        configuration_with_default = self.__force_default_values(
```

Comment:
force default value is already done in parse() why doing it twice?

Reply:
aren't you performing multiple calls to _key?

Also I think the name of the function should be changed